### DISCUSSION TITLE
Linewise jump that preserves current column

### DISCUSSION THREAD

**Discussion by Andrew15-5** [ggandor/leap.nvim#222] [Q&A] [answered]
I've used the provided linewise snippet for a long time and was using default line numbers too. I finally decided to try relative numbers to be able to jump straight up and down in normal/visual modes. I also wonder if that snippet can be modified to achieve the same result. Currently, it always sends the cursor to the first column, which is really annoying when you want to select a block or some other precise movement (you would have to jump/move to the correct column to "fix" that).

**Comment by ggandor** [comment, marked as answer]
> I also wonder if that snippet can be modified to achieve the same result. 

We actually jump to each line, to get the value of the last column, and choose either the original column, or the last one on the given line (if the original is beyond that). Finally, we restore the view.

```lua
    -- Get targets.
    local saved_view = vim.fn.winsaveview()  -- <<
    local curcol = vim.fn.col('.')           -- <<
    local targets = {}
    local lnum = wininfo.topline
    while lnum <= wininfo.botline do
      local fold_end = vim.fn.foldclosedend(lnum)
      -- Skip folded ranges.
      if fold_end ~= -1 then
        lnum = fold_end + 1
      else
        if (lnum < cur_line - skip_range) or (lnum > cur_line + skip_range) then
          vim.fn.cursor(lnum, 0)                                                                 -- <<
          local col = (curcol < vim.fn.col('$')) and curcol or math.max(1, vim.fn.col('$') - 1)  -- <<
          table.insert(targets, { pos = { lnum, col } })                                         -- <<
        end
        lnum = lnum + 1
      end
    end
    vim.fn.winrestview(saved_view)  -- <<
```
This doesn't handle multibyte characters properly though, but can be tweaked further.

I'm thinking about adding a `pattern` parameter to `leap()` (search pattern instead of prepared targets), so that this whole snippet can be replaced with a one-line call...

**Reply by Andrew15-5** [reply]
The `vim.fn.winrestview(saved_view)` line says:

```
1. Cannot assign `vim.fn.winsaveview.ret` to parameter `vim.fn.winrestview.dict`.
   - `vim.fn.winsaveview.ret` cannot match `vim.fn.winrestview.dict`
   - Type `vim.fn.winsaveview.ret` cannot match `vim.fn.winrestview.dict` [param-type-mismatch]
```

But either way it does work now, how I wanted it to work. And since the targets are at the same column as where I'm looking at, it gets even easier. Thanks!

<details><summary>Updated function</summary>

```lua
local function get_line_starts(winid, skip_range)
  local wininfo = vim.fn.getwininfo(winid)[1]
  local cur_line = vim.fn.line "."
  -- Skip lines close to the cursor.
  local skip_range = skip_range or 2

  -- Get targets.
  local saved_view = vim.fn.winsaveview() -- <<
  local curcol = vim.fn.col "." -- <<
  local targets = {}
  local lnum = wininfo.topline
  while lnum <= wininfo.botline do
    local fold_end = vim.fn.foldclosedend(lnum)
    -- Skip folded ranges.
    if fold_end ~= -1 then
      lnum = fold_end + 1
    else
      if (lnum < cur_line - skip_range) or (lnum > cur_line + skip_range) then
        vim.fn.cursor(lnum, 0) -- <<
        local col = (curcol < vim.fn.col "$") and curcol or math.max(1, vim.fn.col "$" - 1) -- <<
        table.insert(targets, { pos = { lnum, col } }) -- <<
      end
      lnum = lnum + 1
    end
  end
  vim.fn.winrestview(saved_view) -- <<

  -- Sort them by vertical screen distance from cursor.
  local cur_screen_row = vim.fn.screenpos(winid, cur_line, 1)["row"]
  local function screen_rows_from_cur(t)
    local t_screen_row = vim.fn.screenpos(winid, t.pos[1], t.pos[2])["row"]
    return math.abs(cur_screen_row - t_screen_row)
  end
  table.sort(targets, function(t1, t2)
    return screen_rows_from_cur(t1) < screen_rows_from_cur(t2)
  end)

  if #targets >= 1 then
    return targets
  end
end
```

</details>

**Reply by Andrew15-5** [reply]
I think now it's kinda a preference thing, since for leap-based up-down jumps you have to invoke it, and then press one or two keys and you are here. For the line number based up-down jumps you have to look at the number on the left side, write a 1 or 2 digit number and press j/k. They are very close, but now with leap you don't have to move your eyes from the target to look at the start of the line.
